File: src/main/java/systems/conduit/main/core/plugin/PluginManager.java
package systems.conduit.main.core.plugin;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.Multimap;
import javassist.tools.Callback;
import lombok.Getter;
import systems.conduit.main.Conduit;
import systems.conduit.main.core.events.types.ServerEvents;
import systems.conduit.main.core.plugin.annotation.Dependency;
import systems.conduit.main.core.plugin.annotation.DependencyType;
import systems.conduit.main.core.plugin.annotation.PluginMeta;

import java.io.File;
import java.io.IOException;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Collectors;

public class PluginManager {

    @Getter private Queue<Plugin> plugins = new ConcurrentLinkedQueue<>();

    /**
     * Finds all classes in runtime that extend {@link Plugin} and are annotated with {@link systems.conduit.main.core.plugin.annotation.PluginMeta}
     * and attempts to load it as a plugin.
     */
    public void loadPlugins() {
        Conduit.getLogger().info("Loading plugins...");
        // We can now get all the jars in the plugins folder and load all of them.
        File pluginsFolder = Paths.get("plugins").toFile();
        if (!pluginsFolder.exists() && !pluginsFolder.mkdirs()) {
            Conduit.getLogger().error("Failed to make plugins directory.");
            return;
        }
        File[] pluginFiles = pluginsFolder.listFiles();
        if (pluginFiles == null) return;
        List<File> plugins = new ArrayList<>();
        for (File file : pluginFiles) {
            // Skip folders, and non-jars
            if (!file.isFile() || !file.getName().endsWith(".jar")) continue;
            // Since it is a file, and it ends with .jar, we can proceed with attempting to load it.
            plugins.add(file);
        }
        loadPlugins(plugins, false);
    }

    private void loadPlugins(List<File> pluginFiles, boolean reload) {
        try {
            // List that should never be changed unless plugin is unable to ever be loaded.
            List<PluginMeta> metas = new ArrayList<>();
            // Loaders - Should be able to be removed from here if we loaded the plugin as plugin now has classloader on it.
            Map<PluginMeta, PluginClassLoader> loaders = new HashMap<>();
            // Needed to be loaded first soft dependencies - Remove plugin from map and any lists
            Multimap<PluginMeta, String> softDependencies = ArrayListMultimap.create();

            for (File pluginFile : pluginFiles) {
                try (PluginClassLoader classLoader = new PluginClassLoader(pluginFile, this.getClass().getClassLoader())) {
                    Optional<PluginMeta> pluginMeta = classLoader.loadMeta();
                    pluginMeta.ifPresent(meta -> loaders.put(meta, classLoader));
                    pluginMeta.ifPresent(metas::add);
                } catch (IOException e) {
                    Conduit.getLogger().error("Error loading plugin: " + pluginFile.getName());
                    e.printStackTrace();
                }
            }
            for (PluginMeta meta : loaders.keySet()) {
                // All soft dependencies
                List<PluginMeta> otherDependencies = metas.stream().filter(m -> !m.equals(meta)).collect(Collectors.toList());
                softDependencies.putAll(meta, getSoftDependencies(otherDependencies, meta));
            }
            while (!loaders.isEmpty()) {
                for (Iterator<Map.Entry<PluginMeta, PluginClassLoader>> it = loaders.entrySet().iterator(); it.hasNext();) {
                    Map.Entry<PluginMeta, PluginClassLoader> pl = it.next();
                    List<String> hardDependencies = getHardDependencies(pl.getKey());
                    // Check to make sure all hard dependencies are able to be loaded
                    if (!getNames(metas).containsAll(getHardDependencies(pl.getKey()))) {
                        Conduit.getLogger().error("Failed to load " + pl.getKey().name() + " due to missing dependencies.");
                        hardDependencies.removeAll(getNames(metas));
                        Conduit.getLogger().error("Required dependencies: " + hardDependencies);
                        // Can't load ever so remove all reference to plugin
                        softDependencies.asMap().remove(pl.getKey());
                        metas.remove(pl.getKey());
                        it.remove();
                        return;
                    }
                    // Load soft soft dependencies first
                    if (softDependencies.get(pl.getKey()).isEmpty()) {
                        Optional<Plugin> optionalPlugin = loaders.get(pl.getKey()).load(pl.getKey());
                        // The plugin should now be loaded, now we can attempt to enable the plugin.
                        if (!optionalPlugin.isPresent()) return;
                        Plugin plugin = optionalPlugin.get();
                        enable(plugin, reload);
                        this.plugins.add(plugin);
                        softDependencies.values().remove(plugin.getMeta().name());
                        softDependencies.asMap().remove(plugin.getMeta());
                        it.remove();
                    }
                }
            }
        } catch (Exception e) {
            Conduit.getLogger().error("Error loading plugin.");
            e.printStackTrace();
        }
    }

    private List<String> getNames(Collection<PluginMeta> metas) {
        return metas.stream().map(PluginMeta::name).collect(Collectors.toList());
    }

    private List<String> getHardDependencies(PluginMeta meta) {
        return Arrays.stream(meta.dependencies()).filter(dependency -> dependency.type() == DependencyType.HARD).map(Dependency::name).collect(Collectors.toList());
    }

    private List<String> getSoftDependencies(Collection<PluginMeta> metas, PluginMeta meta) {
        List<String> names  = getNames(metas);
        return Arrays.stream(meta.dependencies())
                .filter(dependency -> dependency.type() == DependencyType.SOFT).filter(dependency -> names.contains(dependency.name()))
                .map(Dependency::name).collect(Collectors.toList());
    }

    public void disablePlugins() {
        Conduit.getLogger().info("Disabling plugins...");
        if (plugins.isEmpty()) return;
        // Loop through all plugins and disable them
        Iterator<Plugin> iterator = plugins.iterator();
        while (iterator.hasNext()) {
            Plugin plugin = iterator.next();
            // Disable the plugin
            disable(plugin, false);
            // Remove plugin from list
            iterator.remove();
        }
    }

    public void enable(Plugin plugin, boolean reload) {
        // Enable now
        if (!reload) Conduit.getLogger().info("Enabling plugin: " + plugin.getMeta().name());
        plugin.setPluginState(PluginState.ENABLING);
        plugin.onEnable();
        plugin.setPluginState(PluginState.ENABLED);
        if (!reload) Conduit.getLogger().info("Enabled plugin: " + plugin.getMeta().name());
    }

    public void disable(Plugin plugin, boolean server) {
        if ((plugin.getPluginState() == PluginState.DISABLING || plugin.getPluginState() == PluginState.DISABLED)) return;
        if (!server) Conduit.getLogger().info("Disabling plugin: " + plugin.getMeta().name());
        plugin.setPluginState(PluginState.DISABLING);
        plugin.getEvents().clear();
        plugin.onDisable();
        plugin.setPluginState(PluginState.DISABLED);
        if (!server) Conduit.getLogger().info("Disabled plugin: " + plugin.getMeta().name());
    }

    public void reloadPlugins(boolean server, Callback callback) {
        if (!server) Conduit.getLogger().info("Reloading all plugins...");
        plugins.forEach(plugin -> reload(plugin, false));
        callback.result(new Object[]{});
        if (!server) Conduit.getLogger().info("Reloaded all plugins");
    }

    public void reload(Plugin plugin, boolean server) {
        // Before we attempt to reload the plugin, make sure that it can safely be done.
        if (!plugin.getMeta().reloadable()) {
            // This plugin is not reloadable.
            return;
        }
        if (!server) Conduit.getLogger().info("Reloading plugin: " + plugin.getMeta().name());
        // Unload the plugin
        disable(plugin, true);
        AtomicReference<Optional<File>> pluginFile = new AtomicReference<>(Optional.empty());
        // Remove the plugin if found and store the file for later
        plugins.removeIf(pl -> {
            pluginFile.set(Optional.ofNullable(pl.getClassLoader().getPluginFile()));
            return pl.equals(plugin);
        });
        // Load plugin again
        pluginFile.get().ifPresent(file -> loadPlugins(Collections.singletonList(file), true));
        if (!server) Conduit.getLogger().info("Reloaded plugin: " + plugin.getMeta().name());

        ServerEvents.PluginReloadEvent event = new ServerEvents.PluginReloadEvent(plugin);
        Conduit.getEventManager().dispatchEvent(event);
    }

    public Optional<Plugin> getPlugin(String name) {
        return plugins.parallelStream().filter(plugin -> plugin.getMeta().name().equalsIgnoreCase(name)).findFirst();
    }
}

File: src/main/kotlin/systems/conduit/core/plugin/PluginManager.kt
package systems.conduit.core.plugin

import com.google.common.collect.ArrayListMultimap
import com.google.common.collect.Multimap
import javassist.tools.Callback
import systems.conduit.core.Conduit
import systems.conduit.core.plugin.annotation.Dependency
import systems.conduit.core.plugin.annotation.DependencyType
import systems.conduit.core.plugin.annotation.PluginMeta
import java.io.File
import java.io.IOException
import java.nio.file.Paths
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.atomic.AtomicReference
import java.util.function.Consumer
import java.util.function.Function
import java.util.function.Predicate
import java.util.stream.Collectors

class PluginManager {

    var plugins: Queue<Plugin> = ConcurrentLinkedQueue()
        private set

    /**
     * Finds all classes in runtime that extend [Plugin] and are annotated with [systems.conduit.core.plugin.annotation.PluginMeta]
     * and attempts to load it as a plugin.
     */
    fun loadPlugins() {
        Conduit.logger.info("Loading plugins...")
        // We can now get all the jars in the plugins folder and load all of them.
        val pluginsFolder = Paths.get("plugins").toFile()
        if (!pluginsFolder.exists() && !pluginsFolder.mkdirs()) {
            Conduit.logger.error("Failed to make plugins directory.")
            return
        }
        val pluginFiles = pluginsFolder.listFiles() ?: return
        val plugins: MutableList<File> = ArrayList()
        for (file in pluginFiles) {
            // Skip folders, and non-jars
            if (!file.isFile || !file.name.endsWith(".jar")) continue
            // Since it is a file, and it ends with .jar, we can proceed with attempting to load it.
            plugins.add(file)
        }
        loadPlugins(plugins, false)
    }

    private fun loadPlugins(pluginFiles: List<File>, reload: Boolean) {
        try {
            // List that should never be changed unless plugin is unable to ever be loaded.
            val metas: MutableList<PluginMeta> = ArrayList()
            // Loaders - Should be able to be removed from here if we loaded the plugin as plugin now has classloader on it.
            val loaders: MutableMap<PluginMeta?, PluginClassLoader> = HashMap()
            // Needed to be loaded first soft dependencies - Remove plugin from map and any lists
            val softDependencies: Multimap<PluginMeta?, String> = ArrayListMultimap.create()
            for (pluginFile in pluginFiles) {
                try {
                    PluginClassLoader(pluginFile, this.javaClass.classLoader).use { classLoader ->
                        val pluginMeta = classLoader.loadMeta()
                        if (pluginMeta != null) {
                            loaders[pluginMeta] = classLoader
                            metas.add(pluginMeta)
                        }
                    }
                } catch (e: IOException) {
                    Conduit.logger.error("Error loading plugin: " + pluginFile.name)
                    e.printStackTrace()
                }
            }
            for (meta in loaders.keys) {
                // All soft dependencies
                val otherDependencies = metas.stream().filter { m: PluginMeta? -> m != meta }.collect(Collectors.toList())
                softDependencies.putAll(meta, getSoftDependencies(otherDependencies, meta))
            }
            while (loaders.isNotEmpty()) {
                val it: MutableIterator<Map.Entry<PluginMeta?, PluginClassLoader>> = loaders.entries.iterator()
                while (it.hasNext()) {
                    val pl = it.next()
                    val hardDependencies = getHardDependencies(pl.key)
                    // Check to make sure all hard dependencies are able to be loaded
                    if (!getNames(metas).containsAll(getHardDependencies(pl.key))) {
                        Conduit.logger.error("Failed to load " + pl.key!!.name + " due to missing dependencies.")
                        hardDependencies.removeAll(getNames(metas))
                        Conduit.logger.error("Required dependencies: $hardDependencies")
                        // Can't load ever so remove all reference to plugin
                        softDependencies.asMap().remove(pl.key)
                        metas.remove(pl.key)
                        it.remove()
                        return
                    }
                    // Load soft soft dependencies first
                    if (softDependencies[pl.key].isEmpty()) {
                        val optionalPlugin = loaders[pl.key]!!.load(pl.key) ?: return
                        // The plugin should now be loaded, now we can attempt to enable the plugin.
                        enable(optionalPlugin, reload)
                        plugins.add(optionalPlugin)
                        softDependencies.values().remove(optionalPlugin.meta.name())
                        softDependencies.asMap().remove(optionalPlugin.meta)
                        it.remove()
                    }
                }
            }
        } catch (e: Exception) {
            Conduit.logger.error("Failed to load plugin.")
            e.printStackTrace()
        }
    }

    private fun getNames(metas: List<PluginMeta>): List<String?> {
        return metas.map { it.name }.toList()
    }

    private fun getHardDependencies(meta: PluginMeta) = meta.dependencies.filter { it.type == DependencyType.HARD }.map { it.name }
    private fun getSoftDependencies(metas: List<PluginMeta>, meta: PluginMeta) = meta.dependencies.filter { it.type == DependencyType.SOFT }.filter { getNames(metas).contains(it.name) }.map { it.name }

    fun disablePlugins() {
        if (plugins.isEmpty()) return
        Conduit.logger.info("Disabling plugins...")
        // Loop through all plugins and disable them
        val iterator = plugins.iterator()
        while (iterator.hasNext()) {
            val plugin = iterator.next()
            // Disable the plugin
            disable(plugin, false)
            // Remove plugin from list
            iterator.remove()
        }
    }

    fun enable(plugin: Plugin, reload: Boolean) {
        // Enable now
        if (!reload) Conduit.logger.info("Enabling plugin: " + plugin.meta.name())
        plugin.pluginState = PluginState.ENABLING
        plugin.onEnable()
        plugin.pluginState = PluginState.ENABLED
        if (!reload) Conduit.logger.info("Enabled plugin: " + plugin.meta.name())
    }

    fun disable(plugin: Plugin, server: Boolean) {
        if (plugin.pluginState == PluginState.DISABLING || plugin.pluginState == PluginState.DISABLED) return
        if (!server) Conduit.logger.info("Disabling plugin: " + plugin.meta.name())
        plugin.pluginState = PluginState.DISABLING
        plugin.events.clear()
        plugin.onDisable()
        plugin.pluginState = PluginState.DISABLED
        if (!server) Conduit.logger.info("Disabled plugin: " + plugin.meta.name())
    }

    fun reloadPlugins(server: Boolean, callback: Callback) {
        if (!server) Conduit.logger.info("Reloading all plugins...")
        plugins.forEach(Consumer { plugin: Plugin -> reload(plugin, false) })
        callback.result(arrayOf())
        if (!server) Conduit.logger.info("Reloaded all plugins")
    }

    fun reload(plugin: Plugin, server: Boolean) {
        // Before we attempt to reload the plugin, make sure that it can safely be done.
        if (!plugin.meta.reloadable()) {
            // This plugin is not reloadable.
            return
        }
        if (!server) Conduit.logger.info("Reloading plugin: " + plugin.meta.name())
        // Unload the plugin
        disable(plugin, true)
        val pluginFile = AtomicReference(Optional.empty<File>())
        // Remove the plugin if found and store the file for later
        plugins.removeIf { pl: Plugin ->
            pluginFile.set(Optional.ofNullable(pl.getClassLoader().pluginFile))
            pl == plugin
        }
        // Load plugin again
        pluginFile.get().ifPresent { file: File -> loadPlugins(listOf(file), true) }
        if (!server) Conduit.logger.info("Reloaded plugin: " + plugin.meta.name())
        val event = PluginReloadEvent(plugin)
        Conduit.eventManager.dispatchEvent(event)
    }

    fun getPlugin(name: String): Plugin? = plugins.first { it.meta.name().equals(name, ignoreCase = true) }

}
